flowmachine/flowmachine/features/subscriber/mobility_classification.py

```diff
@@ -0,0 +1,103 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from typing import List
+
+from flowmachine.features.subscriber.metaclasses import SubscriberFeature
+from flowmachine.features.utilities.subscriber_locations import BaseLocation
+from flowmachine.features.subscriber.subscriber_stay_lengths import (
+    SubscriberStayLengths,
+)
+from flowmachine.core.errors import InvalidSpatialUnitError
+
+
+class MobilityClassification(SubscriberFeature):
+    """
+    Based on subscribers' reference locations in a sequence of reference
+    periods, classify each subscriber as having one of the following mobility
+    types (the assigned label corresponds to the first of these criteria that
+    is true for a given subscriber):
+
+    - 'unlocated': Subscriber has a NULL location in the most recent period
+    - 'irregular': Subscriber is not active in at least one of the reference
+      periods
+    - 'not_always_locatable': Subscriber has a NULL location in at least one of
+      the reference periods
+    - 'mobile': Subscriber spent fewer than 'stay_length_threshold' consecutive
+      periods at any single location
+    - 'stable': Subscriber spent at least 'stay_length_threshold' consecutive
+      periods at the same location
+
+    Only subscribers appearing in the result of the reference location query
+    for the most recent period are included in the result of this query (i.e.
+    subscribers absent from the query result can be assumed to fall into a
+    sixth category: "not active in the most recent period").
+
+    Parameters
+    ----------
+    locations : list of BaseLocation
+        List of reference location queries, each returning a single location
+        per subscriber (or NULL location for subscribers that are active but
+        unlocatable). The list is assumed to be sorted into ascending
+        chronological order.
```

Comment:
Would it matter if it's not, and if not, is it worth adding a sort to the `subscriber_stay_lengths` query?